compiler/rustc_trait_selection/src/traits/select/mod.rs

```diff
@@ -1666,39 +1665,45 @@ impl<'cx, 'tcx> SelectionContext<'cx, 'tcx> {
     fn match_normalize_trait_ref(
         &mut self,
         obligation: &PolyTraitObligation<'tcx>,
-        trait_bound: ty::PolyTraitRef<'tcx>,
         placeholder_trait_ref: ty::TraitRef<'tcx>,
-    ) -> Result<Option<ty::PolyTraitRef<'tcx>>, ()> {
+        trait_bound: ty::PolyTraitRef<'tcx>,
+    ) -> Result<Option<ty::TraitRef<'tcx>>, ()> {
         debug_assert!(!placeholder_trait_ref.has_escaping_bound_vars());
         if placeholder_trait_ref.def_id != trait_bound.def_id() {
             // Avoid unnecessary normalization
             return Err(());
         }
 
-        let Normalized { value: trait_bound, obligations: _ } = ensure_sufficient_stack(|| {
-            project::normalize_with_depth(
-                self,
-                obligation.param_env,
-                obligation.cause.clone(),
-                obligation.recursion_depth + 1,
+        self.infcx.probe(|_| {
+            let trait_bound = self.infcx.instantiate_binder_with_fresh_vars(
+                obligation.cause.span,
+                HigherRankedType,
                 trait_bound,
-            )
-        });
-        self.infcx
-            .at(&obligation.cause, obligation.param_env)
-            .sup(DefineOpaqueTypes::No, ty::Binder::dummy(placeholder_trait_ref), trait_bound)
-            .map(|InferOk { obligations: _, value: () }| {
-                // This method is called within a probe, so we can't have
-                // inference variables and placeholders escape.
-                if !trait_bound.has_infer() && !trait_bound.has_placeholders() {
-                    Some(trait_bound)
-                } else {
-                    None
-                }
-            })
-            .map_err(|_| ())
+            );
+            let Normalized { value: trait_bound, obligations: _ } = ensure_sufficient_stack(|| {
+                project::normalize_with_depth(
+                    self,
+                    obligation.param_env,
+                    obligation.cause.clone(),
+                    obligation.recursion_depth + 1,
+                    trait_bound,
+                )
+            });
+            self.infcx
+                .at(&obligation.cause, obligation.param_env)
+                .eq(DefineOpaqueTypes::No, placeholder_trait_ref, trait_bound)
+                .map(|InferOk { obligations: _, value: () }| {
+                    // This method is called within a probe, so we can't have
+                    // inference variables and placeholders escape.
+                    if !trait_bound.has_infer() && !trait_bound.has_placeholders() {
```

Comment:
because we instantiate binders with infer vars, we anow lways fail this condition for higher ranked bounds.

However, because higher ranked bounds use different `DefId`s for their bound vars anyways, `for<'a> Something` and `for<'a> Something` are structurally different, so we don't deduplicate them anyways. See https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=77855b4377e6568538caeb13dcd2d4df